app/Models/Concerns/HasMemberships.php

```diff
@@ -0,0 +1,159 @@
+<?php
+
+namespace App\Models\Concerns;
+
+use App\Exceptions\Memberships\MembershipNotSecondaryException;
+use App\Exceptions\Memberships\PrimaryMembershipDoesntAllowSecondaryException;
+use App\Models\Membership;
+use App\Models\Membership\MembershipPivot;
+use Carbon\Carbon;
+use Illuminate\Database\Eloquent\Relations\BelongsToMany;
+
+trait HasMemberships
+{
+    private $primaryMembership;
+
+    public function memberships(): BelongsToMany
+    {
+        return $this->membershipsRelationship()
+            ->wherePivot('ended_at', null)
+            ->orderByPriority();
+    }
+
+    public function membershipHistory(): BelongsToMany
+    {
+        return $this->membershipsRelationship()
+            ->orderBy((new MembershipPivot())->getTable().'.started_at', 'desc');
+    }
+
+    public function primaryMembership(): ?Membership
+    {
+        if ($this->primaryMembership) {
+            return $this->primaryMembership;
+        }
+
+        return $this->primaryMembership = $this->memberships()
+            ->primary()
+            ->first();
+    }
+
+    public function secondaryMemberships(): BelongsToMany
+    {
+        return $this->memberships()
+            ->secondary()
+            ->orderByPriority();
+    }
+
+    public function updatePrimaryMembership(string $division, string $region): bool
```

Comment:
This method has a lot going on and is fairly confusing to follow. Lots of temporary variables, lots of conditionals and then flushing the cache at the end?